enhancements/sig-policy/74-order-policy-execution/README.md

```diff
@@ -0,0 +1,250 @@
+since the policy object for a pending template will not exist on the cluster.
+
+There are some instances where users might set a policy template to only
+activate when the cluster deviates from a "normal" state. For example, an inform
+policy could be set to watch for the existence of a deployment on a cluster, and
+a policy template dependent on that inform policy could re-create the deployment
+if the inform policy enters a NonCompliant state. If the dependent template is
+in a policy with other compliant templates, we do not want that parent policy to have a
+Pending status because the normal state of the pending template is inactive. We
+plan on adding a `spec.policyTemplates[].ignorePending` field so that users can
+ignore if a template is pending while calculating the overall compliance of a policy.
+
+### Risks and Mitigation
+
+- Performance hit due to having to watch extra resources (utilizing the watch
+library should help with this)
+- Policies that are dependent on another policy status to be processed will
+take longer to show status than normal policies. This shouldn't be a big issue
+most of the time, but if the config policy controller is at its saturation point
+(500+ configuration policies) it could take over a minute for the dependent policy
+status to update correctly
+- Circular dependencies: it is possible for users to create a loop where policy A
+has a dependency on policy B, and policy B has a dependency on policy A so they
+will both get stuck in a pending state. We don't plan to check for this since it
+is user error and we expect the status message accompanying a pending policy
+to be detailed enough that a user who has created a circular dependency would
+be able to trace their chain of dependencies to find the issue.
+
+## Design Details
+
+Here is what the new spec field will look like in a policy. The top-level
+`dependencies` field applies to all policy templates in the policy, while
+the template-level `extraDependencies` field can be used to set up
+dependencies for a specific policy template. The top-level `dependencies`
+will take priority over the template-level `extraDependencies`, so policy
+templates that have `extraDependencies` satisfied but not `dependencies`
+will still be marked as `pending`. This means that the dependencies for
+a policy will be the union of its `dependencies` and `extraDependencies`
+fields.
+
+```
+apiVersion: policy.open-cluster-management.io/v1
+kind: Policy
+metadata:
+  name: my-complicated-policy
+spec:
+  dependencies: //for all policy templates
+    - apiVersion: policy.open-cluster-management.io/v1
+      kind: Policy
+      name: namespace-foo-setup-policy
+      namespace: policies
+      compliance: Compliant
+  policy-templates:
+    - extraDependencies: //for this policy template only
+        - apiVersion: policy.open-cluster-management.io/v1
+          kind: ConfigurationPolicy
+          name: bar
+          namespace: local-cluster
```

Comment:
I believe it will be better UX , if we simplify the `extraDependencies` by considering the order of the objectDefinations under the `policy-templates->[]objectDefinitions. `  for ex;

```
policy-templates:
-  objectDefinitions:
     apiVersion: policy.open-cluster-management.io/v1
     kind: ConfigurationPolicy
     metadata:
        name: policy1
      spec:
          .....
- ignorePending: false
  compliance: compliant
  objectDefinitions:
     apiVersion: policy.open-cluster-management.io/v1
     kind: ConfigurationPolicy
     metadata:
        name: policy2
      spec:
          .....

```
This way we avoid re-write the configurationPolicy info each time AND it make sense to have a constrain for having configurationPolicies with dependencies must be defined under the same Policy-Templates.

OR

Another way of considering the order is by define order with numbers. Ex; the policy that creates the NS with order value 0 while all other policies with order value 1. so the order of the []objectDefinations doesn't matter (to avoid back compatibility issue ) something like below 

```
policy-templates:
- policyOrder: 0 (default is 0 so all policies have the same order to apply unless user specify)
  objectDefinitions:
     apiVersion: policy.open-cluster-management.io/v1
     kind: ConfigurationPolicy
     metadata:
        name: policy1
      spec:
          .....
- policyOrder: 1
  ignorePending: false
  compliance: compliant
  objectDefinitions:
     apiVersion: policy.open-cluster-management.io/v1
     kind: ConfigurationPolicy
     metadata:
        name: policy2
      spec:
          .....
- policyOrder: 1
  ignorePending: true
  compliance: compliant
  objectDefinitions:
     apiVersion: policy.open-cluster-management.io/v1
     kind: ConfigurationPolicy
     metadata:
        name: policy3
      spec:
          .....
- policyOrder: 2
  ignorePending: true
  compliance: compliant
  objectDefinitions:
     apiVersion: policy.open-cluster-management.io/v1
     kind: ConfigurationPolicy
     metadata:
        name: policy4
      spec:
          .....
```

Reply:
These suggestions could simplify the case where templates inside a Policy naturally chain together, but I think they have disadvantages in other cases. I would rather have these kinds of options in the policy generator, and have that tool deal with the tedious "wiring" of the details, if possible.

An example of something I'd like to see, but I don't think would work under either of these suggestions:
- template A has no dependencies
- template B should activate when template A is compliant
- template C should activate when template B is compliant
- template D should activate then template A is terminating
```
A -> B -> C
 `-> D
```

I would also rather only have one way of declaring dependencies: these would remove `extraDependencies` inside each template, but would still have the top-level `dependencies` field.

Reply:
It make sense for Policy dependencies to be defined as mentioned in the example as below ;
```
  dependencies: //for all policy templates
    - apiVersion: policy.open-cluster-management.io/v1
      kind: Policy
      name: namespace-foo-setup-policy
      namespace: policies
      compliance: Compliant
```
however for policy-templates->(configurationPolicies) I still believe it can be simplified similar to my previous comment. 
Just to clarify the case as endUser; if i have ConfigPolicy1 defined in Policy-template1 that will be an extraDependency for  configPolicy2 defined in policy-template2. I would level up policy-template1 to be dependency for policy-template2.

AND agree policyGenerator will definitely help a lot on simplifying the policy creation.